sources/api/apiclient/src/exec.rs

```diff
@@ -0,0 +1,683 @@
+/// size) and alerting the caller for terminal signals.
+struct HandleSignals {
+    /// If a terminal signal is received, its value is sent over this channel.
+    signal_rx: oneshot::Receiver<i32>,
+}
+
+impl HandleSignals {
+    /// Parameters:
+    /// * winch_tx: The channel to which we should send window size change messages.
+    fn new(winch_tx: mpsc::UnboundedSender<Message>) -> Result<Self> {
+        // Create a channel we use to tell the caller when we receive a terminal signal.
+        let (signal_tx, signal_rx) = oneshot::channel();
+
+        // Set up the signal handler; do this before starting a thread so we can die quickly on
+        // failure.
+        use signal::*;
+        let signals =
+            Signals::new(&[SIGWINCH, SIGTERM, SIGINT, SIGQUIT]).context(error::HandleSignals)?;
+
+        debug!("Spawning thread to manage signals");
+        thread::spawn(move || {
+            if let Err(e) = Self::handle_signals(signals, winch_tx, signal_tx) {
+                error!("Signal manager failed: {}", e);
+            }
+        });
+
+        Ok(Self { signal_rx })
+    }
+
+    fn handle_signals(
+        mut signals: Signals,
+        winch_tx: mpsc::UnboundedSender<Message>,
+        signal_tx: oneshot::Sender<i32>,
+    ) -> Result<()> {
+        use signal::*;
+        loop {
+            // Block until our process receives a signal.
+            for signal in signals.wait() {
+                if signal == SIGWINCH {
+                    // Window size changes can happen any number of times; send an update to the
+                    // server and wait for more signals.
+                    Self::send_winch(&winch_tx);
+                } else {
+                    // Anything else is terminal; notify the caller and exit.
+                    signal_tx
+                        .send(signal)
+                        .ok()
+                        .context(error::SendSignal { signal })?;
+                    // The signal and our handler have done their job, it's not an error.
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    /// Try to send a window size update to the server.  We don't consider window size updates to
+    /// be critical, since the program is still functioning, so we don't return errors.
+    fn send_winch(tx: &mpsc::UnboundedSender<Message>) {
+        if let Some(winsize) = get_winsize(STDOUT_FILENO) {
+            debug!(
+                "Sending new window size to server: {} cols {} rows",
+                winsize.cols, winsize.rows
+            );
+            if let Ok(msg) = serde_json::to_string(&ClientMessage::Winch(winsize)) {
+                let _ = tx.unbounded_send(Message::Text(msg));
+            }
+        }
+    }
+}
+
+/// Get the current window size of the user's terminal, if possible.  We don't consider window size
+/// to be critical, since the program is still functioning, so we return Option rather than Result.
```

Comment:
nit: this comment seems a bit redundant with the comment for `send_winch` just above it

Reply:
I tried to write from the perspective of a reader coming in from different angles; maybe they saw the call to get_winsize in Terminal and go to its definition, or maybe they're trying to understand HandleSignals and don't care about the details of getting the size.  They're also slightly different in behavior and return value, which affects usage.